src/coreclr/jit/loopcloning.cpp

```diff
@@ -2135,18 +2135,7 @@ void Compiler::optCloneLoop(FlowGraphNaturalLoop* loop, LoopCloneContext* contex
         newBlk->scaleBBWeight(slowPathWeightScaleFactor);
         blk->scaleBBWeight(fastPathWeightScaleFactor);
 
-// TODO: scale the pred edges of `blk`?
-
-#if FEATURE_LOOP_ALIGN
-        // If the original loop is aligned, do not align the cloned loop because cloned loop will be executed in
-        // rare scenario. Additionally, having to align cloned loop will force us to disable some VEX prefix encoding
-        // and adding compensation for over-estimated instructions.
-        if (newBlk->isLoopAlign())
```

Comment:
I think we don't clone aligned loops because they are given a very low `bbWeight` by loop cloning, so `optIdentifyLoopsForAlignment` knows that it isn't profitable.

Reply:
I did an ad hoc verification like this:
```diff
diff --git a/src/coreclr/jit/block.h b/src/coreclr/jit/block.h
index ae480701066..b854226a03a 100644
--- a/src/coreclr/jit/block.h
+++ b/src/coreclr/jit/block.h
@@ -431,6 +431,7 @@ enum BasicBlockFlags : unsigned __int64
                                                           // This is just to reduce diffs from removing BBJ_NONE.
                                                           // (TODO: Remove this quirk after refactoring Compiler::fgFindInsertPoint)
     BBF_OLD_LOOP_HEADER_QUIRK          = MAKE_BBFLAG(42), // Block was the header ('entry') of a loop recognized by old loop finding
+    BBF_CLONED_LOOP_HEADER             = MAKE_BBFLAG(43),
 
     // The following are sets of flags.
 
diff --git a/src/coreclr/jit/compiler.cpp b/src/coreclr/jit/compiler.cpp
index 853aa203037..b137e4348e8 100644
--- a/src/coreclr/jit/compiler.cpp
+++ b/src/coreclr/jit/compiler.cpp
@@ -5366,6 +5366,7 @@ void Compiler::optIdentifyLoopsForAlignment(FlowGraphNaturalLoops* loops, BlockT
             loopAlignCandidates++;
             assert(!top->isLoopAlign());
             top->SetFlags(BBF_LOOP_ALIGN);
+            assert(!loop->GetHeader()->HasFlag(BBF_CLONED_LOOP_HEADER));
             JITDUMP("Aligning " FMT_LP " that starts at " FMT_BB ", weight=" FMT_WT " >= " FMT_WT ".\n",
                     loop->GetIndex(), top->bbNum, topWeight, compareWeight);
         }
diff --git a/src/coreclr/jit/loopcloning.cpp b/src/coreclr/jit/loopcloning.cpp
index 36ba5609cb7..5002f69acd5 100644
--- a/src/coreclr/jit/loopcloning.cpp
+++ b/src/coreclr/jit/loopcloning.cpp
@@ -2197,6 +2197,11 @@ void Compiler::optCloneLoop(FlowGraphNaturalLoop* loop, LoopCloneContext* contex
         return BasicBlockVisit::Continue;
     });
 
+    BasicBlock* newHeader = nullptr;
+    bool result = blockMap->Lookup(loop->GetHeader(), &newHeader);
+    assert(result);
+    newHeader->SetFlags(BBF_CLONED_LOOP_HEADER);
+
     // Perform the static optimizations on the fast path.
     optPerformStaticOptimizations(loop, context DEBUGARG(true));
 
diff --git a/src/coreclr/scripts/superpmi.py b/src/coreclr/scripts/superpmi.py
index a06461be609..04b41614562 100644
--- a/src/coreclr/scripts/superpmi.py
+++ b/src/coreclr/scripts/superpmi.py
@@ -1869,7 +1869,7 @@ class SuperPMIReplayAsmDiffs:
 
         # These vars are force overridden in the SPMI runs for both the base and diff, always.
         replay_vars = {
-            "DOTNET_JitAlignLoops": "0", # disable loop alignment to filter noise
+            #"DOTNET_JitAlignLoops": "0", # disable loop alignment to filter noise
             "DOTNET_JitEnableNoWayAssert": "1",
             "DOTNET_JitNoForceFallback": "1",
         }
```
I hit the assert 2 times in coreclr_tests, but nowhere else.
I think the most reasonable "fix" would be to adjust this code in loop cloning:
https://github.com/dotnet/runtime/blob/939f28ababb954ec6966cde6b2eb0b96db40cf69/src/coreclr/jit/loopcloning.cpp#L2021-L2026

It should perhaps be even rarer than 1/100, or mark the cold loop as "run rarely" like the comment says... but I think we can consider it as a follow-up (or leave it as is, since 2 cases in coreclr_tests probably doesn't matter)

Reply:
I'll check the rest of the collections locally. I'd prefer not to have to go through two rounds of CI again for this... I do not think loop alignment should concern itself with how loop cloning created loops -- it should be loop cloning's responsibility to properly mark them as cold.

Reply:
> I'll check the rest of the collections locally

sure, let me know if you find out other failures.

> it should be loop cloning's responsibility to properly mark them as cold.

yes, that's how it should be ideally, but don't think we do it today and the fact that you hit 2 asserts confirms it.

Reply:
> yes, that's how it should be ideally, but don't think we do it today and the fact that you hit 2 asserts confirms it.

Sure -- but I don't think there is a requirement that 100% of cloned loops aren't aligned. I assume the problem in the past was that loop cloning would copy `bbFlags` directly, which already had `BBF_LOOP_ALIGN` set in it. Then we would end up aligning for all cold loops if the hot loop counterpart was aligned. But with the new approach there is a separate profitability calculation for both loops. If we think that the weighing in loop cloning is not right then sure, we should fix it, but I don't think it's a problem we need to fix as part of this PR.

I'll let you know the replay results once I have them.

Reply:
Results: https://gist.github.com/jakobbotsch/03209c3e80fc9c6715a8a790fe8c697a

Looks like the same for all collections except win-x86 has 4 coreclr_tests contexts that hit the assert (it looks like the two extra ones are just duplicate contexts of the other two, not sure why they're only duplicated in the win-x86 collection)

Reply:
> But with the new approach there is a separate profitability calculation for both loops

That makes sense and thanks for running the replay.